File: bms/go/geo/geo_s2_CapFromCenterHeight/geo_s2_CapFromCenterHeight.go
package test

const (
	geo_s1_StraightChordAngle = geo_s1_ChordAngle(4)
	geo_s1_maxLength2         = 4.0
)

func geo_s2_CapFromCenterHeight(center geo_s2_Point, height float64) geo_s2_Cap {
	return geo_s2_CapFromCenterChordAngle(center, geo_s1_ChordAngleFromSquaredLength(2*height))
}

func geo_s1_ChordAngleFromSquaredLength(length2 float64) geo_s1_ChordAngle {
	if length2 > geo_s1_maxLength2 {
		return geo_s1_StraightChordAngle
	}
	return geo_s1_ChordAngle(length2)
}

func geo_s2_CapFromCenterChordAngle(center geo_s2_Point, radius geo_s1_ChordAngle) geo_s2_Cap {
	return geo_s2_Cap{
		center: center,
		radius: radius,
	}
}

type geo_s1_Angle float64

type geo_s1_ChordAngle float64

type geo_s2_Cap struct {
	center geo_s2_Point
	radius geo_s1_ChordAngle
}

type geo_s2_Point struct {
	geo_r3_Vector
}

type geo_r3_Vector struct {
}

File: translations_only/success/go/geo/geo_s2_CapFromCenterHeight/gpt4/CAPR/geo_s2_CapFromCenterHeight.rs

const GEO_S1_STRAIGHT_CHORD_ANGLE: GeoS1ChordAngle = GeoS1ChordAngle(4.0);
const GEO_S1_MAX_LENGTH2: f64 = 4.0;

struct GeoS2Point {
    geo_r3_vector: GeoR3Vector,
}

struct GeoR3Vector {}

struct GeoS2Cap {
    center: GeoS2Point,
    radius: GeoS1ChordAngle,
}

struct GeoS1ChordAngle(f64);

fn geo_s2_cap_from_center_height(center: GeoS2Point, height: f64) -> GeoS2Cap {
    geo_s2_cap_from_center_chord_angle(center, geo_s1_chord_angle_from_squared_length(2.0 * height))
}

fn geo_s1_chord_angle_from_squared_length(length2: f64) -> GeoS1ChordAngle {
    if length2 > GEO_S1_MAX_LENGTH2 {
        return GEO_S1_STRAIGHT_CHORD_ANGLE;
    }
    GeoS1ChordAngle(length2)
}

fn geo_s2_cap_from_center_chord_angle(center: GeoS2Point, radius: GeoS1ChordAngle) -> GeoS2Cap {
    GeoS2Cap { center, radius }
}
